repository/src/main/java/org/apache/atlas/tasks/TaskExecutor.java

```diff
@@ -40,35 +42,29 @@ public class TaskExecutor {
 
     private static final boolean perfEnabled = AtlasPerfTracer.isPerfTraceEnabled(PERF_LOG);
 
-    private final TaskRegistry              registry;
-    private final Map<String, TaskFactory>  taskTypeFactoryMap;
-    private final TaskManagement.Statistics statistics;
-    private final ExecutorService           executorService;
+    private final TaskQueueWatcher watcher;
+
+    static CountDownLatch latch;
 
     public TaskExecutor(TaskRegistry registry, Map<String, TaskFactory> taskTypeFactoryMap, TaskManagement.Statistics statistics) {
-        this.registry           = registry;
-        this.taskTypeFactoryMap = taskTypeFactoryMap;
-        this.statistics         = statistics;
-        this.executorService    = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+        ExecutorService taskExecutorService = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
                                                                     .setDaemon(true)
                                                                     .setNameFormat(TASK_NAME_FORMAT + Thread.currentThread().getName())
                                                                     .build());
-    }
 
-    public void addAll(List<AtlasTask> tasks) {
-        for (AtlasTask task : tasks) {
-            if (task == null) {
-                continue;
-            }
-
-            TASK_LOG.log(task);
+        latch = new CountDownLatch(0);
```

Comment:
Please explain the role of CountDownLatch briefly?

Reply:
Consider it like a counter for number of tasks left in Queue
Whenever loading tasks into the Queue, it is initialized with total number of tasks that are being out into the Queue
As soon as a task is completed it is decremented.

TaskQueueWatcher holds this latch & whenever polling interval is completed it first checks if latch count is 0
if not latch.await() causes TaskQueueWatcher to wait for latch to be 0 & TaskQueueWatcher thread until then